src/internal/coredb/projects.go

```diff
@@ -0,0 +1,171 @@
+package coredb
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/gogo/protobuf/types"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+	"github.com/pachyderm/pachyderm/v2/src/internal/stream"
+	"github.com/pachyderm/pachyderm/v2/src/pfs"
+)
+
+// ProjectIterator batches a page of projectRow entries. Entries can be retrieved using iter.Next().
+type ProjectIterator struct {
+	limit    int
+	offset   int
+	projects []projectRow
+	index    int
+	tx       *pachsql.Tx
+}
+
+// ID is the auto-incrementing primary key used for entries in postgres tables.
+type ID uint64
+
+type projectRow struct {
+	ID          ID        `db:"id"`
+	Name        string    `db:"name"`
+	Description string    `db:"description"`
+	CreatedAt   time.Time `db:"created_at"`
+	UpdatedAt   time.Time `db:"updated_at"`
+}
+
+// Next advances the iterator by one row. It returns a stream.EOS when there are no more entries.
+func (iter *ProjectIterator) Next(ctx context.Context, dst **pfs.ProjectInfo) error {
```

Comment:
How come `Next` writes a pointer to a pointer pointer and returns an error rather than returning a pointer and an error?

If it were writing data into a pointer, that would make sense as a performance hack (a single allocation of the underlying data for the entire run), but this allocations each time through. In languages with only a single return value, you have to take out-args in order to return a value and an error, but in Go you can just return both a value and an error directly.

Reply:
Its to satisfy the stream.ForNext() pattern which uses generics. So if we want to update a *pfs.ProjectInfo, we need a pointer to that pointer in order to update the address of *pfs.ProjectInfo. Perhaps @brendoncarroll can clarify some of the details

Reply:
This is the `stream` API.  `Iterator[*pfs.ProjectInfo]` is going to have a `Next` function that is `**pfs.ProjectInfo`.  Iterator is over `*pfs.ProjectInfo` because it's the type that implements `ProtoMessage()`, not `pfs.ProjectInfo`.

Reply:
Yeah this is just what happens when you mix the `stream.Iterator` with the protobuf types.  It's not ideal, but there are functions like `stream.Next` to make it a little less weird.